### PR TITLE
Proposal to unlocking blind utxo

### PR DESCRIPTION
**Pull request by crisdut**
In addition to the changes in #195, I decided to make a change in `consume_transfer` to unlocking _utxo blinding_.

### CODE REVIEW

**Comment by crisdut** [comment]
Hi @zoedberg , I found the bug.

I did not save the new transition in _transitions_ and _transition_txid_ tables. Also, I added the **CloseMethod** property at the `Reveal` object.

Contract state before spend asset:

image

Contract state after spend asset:

image

I was able to spend an asset received, and I used the change address twice.

I made the tests with branch bellow (I merged the `fix/process_consignment` branch with `feat/consume-reveal`)
* https://github.com/crisdut/rgb-node/tree/exp/prs

PS: I tested only tapret dbc. I will start tests with opret (I never used this).

**Comment by crisdut** [comment]
@dr-orlovsky this branch only merged after merge the #195 branch.

Thanks

File: rpc/src/client.rs [hunk 1]
```diff
@@ -244,10 +245,12 @@ impl Client {
         &mut self,
         transfer: StateTransfer,
         force: bool,
+        reveal: Option<Reveal>,
         progress: impl Fn(String),
     ) -> Result<ContractValidity, Error> {
         self.request(RpcMsg::ConsumeTransfer(AcceptReq {
             consignment: transfer,
+            reveal: reveal,
```
Comment (cryptoquick) [inline rpc/src/client.rs:253]: This line can just be `reveal`, not `reveal: reveal`

Comment (cryptoquick) [inline rpc/src/client.rs:253]: This line can just be `reveal`, not `reveal: reveal`

File: rpc/src/reveal.rs [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+impl Clone for Reveal {
+    fn clone(&self) -> Self {
+        Self {
+            blinding_factor: self.blinding_factor.clone(),
+            outpoint: self.outpoint.clone(),
+            close_method: self.close_method.clone(),
+        }
+    }
+}
```
Comment (cryptoquick) [inline rpc/src/reveal.rs:49]: I can understand the custom Display impl, but can't you just derive Clone and Debug on the `Reveal` struct?

File: rpc/src/reveal.rs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+        let colon_blind = find_blind.unwrap();
+        println!("{:?}", colon_blind);        
+        if colon_blind == 0 || colon_blind == s.len() - 1 {
+            return Err(ParseRevealError::Format);
+        }
+
+        println!("{:?}", &s[..colon_method]);
+        println!("{:?}", &s[colon_method + 1..colon_blind]);
+        println!("{:?}", &s[colon_blind + 1..]);
```
Comment (cryptoquick) [inline rpc/src/reveal.rs:91]: Should these all be printed, or just debug logs?

File: src/bucketd/processor.rs [hunk 4]
```diff
@@ -192,6 +198,37 @@ impl Runtime {
...
+                close_method,
+            }) => {
+                let reveal_outpoint = Revealed {
+                    method: close_method,
+                    blinding: blinding_factor,
+                    txid: Some(outpoint.txid),
+                    vout: outpoint.vout as u32,
+                };
+
+                let conseals = consignment
```
Comment (cryptoquick) [inline src/bucketd/processor.rs:214]: Ah yes, a `conseal`, a concealed seal... :)

**Review by cryptoquick — CHANGES_REQUESTED** [review]
Just some nits, also, can you create steps in the [demo-rgb.sh](https://github.com/LNP-BP/nodes/blob/master/contrib/demo-rgb.sh) to test this as a form of smoke test?

**Comment by zoedberg** [comment]
thanks @crisdut, I confirm you solved the issue I reported and now all rgb-lib tests are passing :)

**Comment by crisdut** [comment]
> Just some nits, also, can you create steps in the demo-rgb.sh to test this as a form of smoke test?

Thanks for reviewing @cryptoquick!

I will start writing the steps, and I'll let you know when I finish them.

**Comment by crisdut** [comment]
Hi @cryptoquick,

The instructions: https://github.com/crisdut/nodes/tree/feat/unlocking-blind-utxo

**Review by cryptoquick — APPROVED** [review]
Just reviewed your changes. Looks good!

**Review by cryptoquick — CHANGES_REQUESTED** [review]
Just realized something, the lints are failing.
Be sure to run: `cargo clippy`

**Comment by dr-orlovsky** [comment]
I have merged #195, so I assume this needs rebase.

Regarding the question from there, which has caused this PR: why would we need two methods instead of one? Well, there were a clear reason, since I have started with just one method and it didn't worked that way, so I had to made two. I am currently sick and my head is not working properly, so I can't recall what was the problem that caused splitting method in two. It could be that it was related to multisig wallets and interactivity required for managing stash under multisig wallets, such that all parties should do one of these two procedures before any of them would proceed to the second one - or something like that. I will try to re-analyse and recall better once I will recover.

**Comment by crisdut** [comment]
> I have merged https://github.com/RGB-WG/rgb-node/pull/195, so I assume this needs rebase.

Yes, you are correct.

> Regarding the question from there, which has caused this PR: why would we need two methods instead of one? 

I my not sure about this question, because I followed the recommendations about extended the method `rgb-cli consume` instead create a new method to reveal the confidential transaction.

**Comment by dr-orlovsky** [comment]
You got me wrong: it was not my question but a quote of Zoe question from here https://github.com/RGB-WG/rgb-node/pull/195#issuecomment-1243346370

And I tried to answer that

**Comment by crisdut** [comment]
Hi @dr-orlovsky,

I fix the PR.

Can you review this code, please?

File: src/bucketd/processor.rs [hunk 5]
```diff
@@ -193,6 +199,37 @@ impl Runtime {
...
+
+                let concealed_seals = consignment
+                    .endpoints()
+                    .filter(|&&(_, seal)| {
+                        reveal_outpoint.to_concealed_seal() == seal.commit_conceal()
+                    })
+                    .clone();
+
+                if concealed_seals.count() == 0 {
+                    eprintln!(
```
Comment (dr-orlovsky) [inline src/bucketd/processor.rs:223]: We must not output to the STDERR - please use `error!` logging instead

File: src/bucketd/processor.rs [hunk 6]
```diff
@@ -193,6 +199,37 @@ impl Runtime {
             self.store.store_sten(db::SCHEMATA, root_schema.schema_id(), root_schema)?;
         }
 
+        match reveal {
+            Some(Reveal {
+                blinding_factor,
+                outpoint,
+                close_method,
+            }) => {
```
Comment (dr-orlovsky) [inline src/bucketd/processor.rs:207]: This is usually done by using `matches!` macro, which avoids the last `_ => none!()` thing. Usually this is reported by the linter - no idea why it passed here, probably linting CI got broken

**Review by dr-orlovsky — COMMENTED** [review]
My initial undetstanding was that RGB Node has to keep the blind UTXO secrets it had generated before and use them internally not requiring the client to provide. I do not remember at this moment does the blind UTXO created by the node - or using RGB libs on a wallet side. If the first is correct, than I would have to re-work this PR and complete that workflow. If not, this PR is overall correct - excluding few points I address in the code review comments

**Comment by crisdut** [comment]
> My initial undetstanding was that RGB Node has to keep the blind UTXO secrets it had generated before and use them internally not requiring the client to provide. I do not remember at this moment does the blind UTXO created by the node - or using RGB libs on a wallet side.

Currently, the blind utxo is created with `rgb-std` lib. I think this process is moved to `rgb-std` after version v0.8.0 of the `rgb-node`.

> If not, this PR is overall correct - excluding few points I address in the code review comments

Great! I will try fix it ASAP. Thanks!

**Review by cryptoquick — APPROVED** [review]
Looks like all the linter checks have passed. Clippy is super helpful. LGTM!

**Review by dr-orlovsky — APPROVED** [review]
utACK 2954e697182468a82ef08d3f9d4d62c28d624aa0

**Pull request merged**